### DISCUSSION TITLE
Step 4. Plugin Development Basic

### DISCUSSION THREAD

**Discussion by choonho** [cloudforet-io/community#139] [Mentorship]
# Plugin Development Basic

## Discord channel for Mentorship

Permanent invite link. https://discord.gg/csp3yt6SWw

# Agenda
  * [Installation of Cloudforet with minikube](https://cloudforet.io/docs/setup_operation/quick_install/)
  * [Fork plugin-http-file-cost-datasource](https://github.com/cloudforet-io/plugin-http-file-cost-datasource)

You can install the Cloudforet, then create a user for service. You can create API user then create API Key for CLI.
You can register the local plugin marketplace for yourself. Then you can enable your custom plugin.

## 1. Installation of Cloudforet

Install Cloudforet following https://cloudforet.io/docs/setup_operation/quick_install/
In the ***Step 6. Set the Helm Values and Upgrade the Chart***, use following values.yaml 

upgrade the Cloudforet with below helm values.yaml file.

* update your ENDPOINT
* update TOKEN 

~~~
console:
  production_json:
    CONSOLE_API:
      ENDPOINT: http://127.0.0.1:8081
    CONSOLE_API_V2:
      ENDPOINT: http://127.0.0.1:8082

cost-analysis:
  enabled: true
  scheduler: true
  worker: true
  replicas_worker: 1

supervisor:
  enabled: true
  image:
    name: cloudforet/supervisor
    version: 1.12.0
  application_scheduler:
    NAME: root
    HOSTNAME: spaceone-plugin.svc.cluster.local
    BACKEND: KubernetesConnector
    CONNECTORS:
      RepositoryConnector:
        endpoint:
          v1: grpc://repository:50051
      PluginConnector:
        endpoint:
          v1: grpc://plugin:50051
      KubernetesConnector:
        namespace: spaceone-plugin
        start_port: 50051
        end_port: 50052
        headless: true
        replica:
          inventory.Collector: 1
          monitoring.DataSource: 1
          cost_analysis.DataSource: 1

global:
  shared_conf:
    TOKEN: ......
~~~

## 2. Create User

Click ***Sign in for Root Account***
![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/f2fe3dee-9e34-43b7-9513-bf557d754ab9)

Login as admin / Admin123!@#

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/734b5d64-f5d0-46a3-84ab-8369ecd43c1a)

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/bcd35c8c-6f79-4974-b50c-e091b49019ed)

Menu: **Administration > IAM**
![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/bb29a039-e89b-44fd-81db-736669da2f9d)

Click ***+ Add***, then create User as ***Domain Admin***

* User ID format is e-mail
* **Check ID** is required
* Set your password
* Enable ***Admin Role***

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/05a153b4-4340-4f2f-9a12-f7e2e6f1639b)

Click profile in upper right. ***Sign Out*** Root Account
![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/79730600-2488-4f06-a7ad-11b98610c749)

Login as general User

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/33b137e6-c03c-46e9-b216-050d3b067e36)

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/c3daffab-e552-40f2-a02a-fe7d849e9290)

## 3. Create API Token

Menu: **Administration > IAM > User**

Check User ID, then select **API Key** in the tap menu.

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/bb344336-1f10-460f-95a3-f79ad270c070)

**+ Create API Key**

***Download YAML*** file
![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/33a32b0e-af7e-4ac8-840c-3b44147504a4)

In the Download directory, there will be ***spacectl_config.yml*** file
![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/0fff7aee-dcc1-41e3-b7c7-cb0758bf425f)

## 4. Access to CLI POD

Cloudforet provides CLI tool, ***spacectl*** as K8S POD

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/a6af0190-2266-46bc-a89e-8c14e8b45b4f)

Access to CLI POD

***kubectl exec -ti {POD NAME} -n spaceone -- /bin/bash***

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/0acfb350-bcbb-4c9f-9d8c-26e5f3dffd67)

In the POD, there is ***default.yml***

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/ed92bae6-a0a6-473a-959e-7c646260c761)

~~~
cp default.yml local.yml
~~~

> default.yml is read only file, you cannot edit default.yml

Update ***local.yml***

* append ***api_key:*** section, using ***spacectl_config.yml*** file in your download directory.

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/a6f73efe-b877-4ef6-8627-0b0bf8a344d2)

Update command

~~~
spacectl config init -f local.yml
~~~

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/7007301b-6588-4ffc-ad51-f948589b6240)

Check API is working

~~~
spacectl api-resources
~~~

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/515dbfa7-7692-4adb-bf9e-0846e3cb4930)

## 5. Create Personal Cloudforet Plugin Repository

Using CLI, create Personal Plugin Repository.

~~~
spacectl exec register repository.Repository -p name="My Private Repo" -p repository_type=local 
~~~
![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/2e2a517b-5756-48e2-89b2-cbba7161c075)

Check the list of repository

~~~
spacectl list repo
~~~

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/53ce12ad-d4f5-4294-b14c-3cc9a6c98e1c)

## 6. Register my own plugin

create ***register_plugin.yml*** file

~~~
name: HTTP file Cost Analysis Data Source
service_type: cost_analysis.DataSource
image: spaceone/plugin-http-file-cost-datasource
capability: {}
tags:
  description: HTTP File Data Source
  icon: https://spaceone-custom-assets.s3.ap-northeast-2.amazonaws.com/console-assets/icons/csv_icon.png
  long_description: This plugin can collect predefined CSV format data through http(s) and use it as cost data. Easily integrate data into Cost Analysis from CSV file regardless of any data as possible like API usage.
template: {}
~~~

***Command***

~~~
spacectl exec register repository.Plugin -f register_plugin.yml
~~~

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/c7422960-efc0-4e7f-9dc1-0e405034a8c7)

Check the plugins

~~~
# List repository
spacectl list repo

# List plugin of a specific repo
spacectl list plugin -p repository_id=<Repository ID>
~~~

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/41b10ba2-f0e9-44f6-9e8f-27bcb0ec6d71)

## 7. Enable Plugin

In your brower, there is default screen for ***Cost Explorer*** menu.
![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/f3b282d7-5b33-4fbf-a9f1-a4ef4a6d0d23)

When you enable ***plugin***, then this screen will be changed to real data screen.

open ***data_source.yml***

~~~
---
name: CSV File
data_source_type: EXTERNAL
plugin_info:
  secret_data:
    status: NOT IMPLEMENTED
  plugin_id: plugin-http-file-cost-datasource
  options:
    base_url:
      - https://raw.githubusercontent.com/cloudforet-io/plugin-http-file-cost-datasource/master/examples/cost_example.csv
    field_mapper:
      additional_info:
        Account: account
~~~

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/c4742346-d89d-4c9f-9953-b3e09f65086c)

Command
~~~
spacectl exec register cost_analysis.DataSource -f data_source.yml
~~~

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/1b8a01ac-ed3c-4839-af34-a925bc024cac)

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/7489f1ed-0f95-45c4-8a09-cf5d3269c22b)

## 8. Sync Cost Data

Remember your data_source_id in Step 7.

~~~
spacectl exec sync cost_analysis.DataSource -p data_source_id=<DATA SOURCE ID> -p start=2023-09
~~~

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/bd8c1e6d-a851-40a9-8d54-d08c9c551796)

![image](https://github.com/cloudforet-io/cloudforet/assets/3993088/4fc85e40-f1da-4484-bbe2-8904288a1bd3)

# Submission
  * Screenshot of your Custom Billing Dashboard

# Reference

spacectl: https://github.com/cloudforet-io/spacectl

**Comment by Cvr421** [comment]
Hello @choonho i have completed the step 4 but i am not able to upload the image on LFX. Is that we should have to upload the pdf file 


![Screenshot from 2024-01-25 18-19-29](https://github.com/cloudforet-io/cloudforet/assets/82499697/f1525fe1-38a6-4f1a-817f-8908c0803daf)

**Reply by choonho** [reply]
Great! You did very well. Don't you have any problem of the minikube memory? I think 4096MB of minikube may be small.

**Reply by Cvr421** [reply]
@choonho 
No actually everything works very smoothly didn't get any Minikube memory issues .
Should i convert the image in pdf format and then upload it to the LFX task.

**Reply by choonho** [reply]
No, I updated your status as completed.

**Reply by Cvr421** [reply]
> No, I updated your status as completed.

Great thanks 👍

**Comment by py-dev-nandini-12** [comment]
Hi @choonho ,
I  have uploaded the task.
<img width="1680" alt="Screenshot 2024-02-01 at 16 27 36" src="https://github.com/cloudforet-io/cloudforet/assets/127988967/edda9e31-06a0-4697-aeff-0bfe756b5d5f">

**Comment by Ayush-Sharma410** [comment]
Hi @choonho 
I have uploaded the task.

**Comment by nileshgadgi** [comment]
@choonho , What is the usecase of this if i have to put the csv file of cost manually, i needed something that will fetch the cost data from the collector and show it in the dashboard for all my projects. Is that feature available by any chance?

**Comment by tthtlc** [comment]
Hi can someone tell me how to register a new plugin?   the command above seemed to be giving error:

my YAML file for the plugin registration (/tmp/plugin-registration3.yaml):

---
capability:
  supported_schema:
  - aws_access_key
image: tthtlc/plugin-hello-world-go:latest
labels:
  - Hello World
  - Demo
  - Go
name: plugin-hello-world-go
plugin_id: plugin-hello-world-go
provider: tthtlc
registry_config:
  url: registry.hub.docker.com
registry_type: DOCKER_HUB
resource_type: inventory.Collector
tags:
  description: Hello World Cloudforet plugin written in Go
  icon: https://spaceone-custom-assets.s3.ap-northeast-2.amazonaws.com/console-assets/icons/default-icon.svg
  
  
  and the command for spacectl registration (inside the spacectl pod):
  
  spacectl exec register repository.Plugin -f /tmp/plugin-registration3.yaml
  
 
  And the errors I got:
  
  ERROR: Required parameter. (key = domain_id)
  
  (i can execute "spacectl exec list repository.Plugin" and "spacectl exec get repository.Plugin -p plugin_id=xxx" without any errors)

**Comment by tthtlc** [comment]
Thanks. I solved it. 🙏🙏🙏
